src/lib.rs

```diff
@@ -3,27 +3,1109 @@
+        amount: NegativeImbalance<T>,
+        remaining_stake: BalanceOf<T>,
+    ) -> NegativeImbalance<T> {
+        let remaining = X::slashed(id, slash_id, amount, remaining_stake);
+        Y::slashed(id, slash_id, remaining, remaining_stake)
+    }
+}
+
+#[derive(Encode, Decode, Copy, Clone, Debug, Default, Eq, PartialEq)]
+pub struct Slash<BlockNumber, Balance> {
+    /// The block where slashing was initiated.
+    pub started_at_block: BlockNumber,
+
+    /// Whether slashing is in active, or conversley paused state.
+    /// Blocks are only counted towards slashing execution delay when active.
+    pub is_active: bool,
+
+    /// The number blocks which must be finalised while in the active period before the slashing can be executed
+    pub blocks_remaining_in_active_period_for_slashing: BlockNumber,
+
+    /// Amount to slash
+    pub slash_amount: Balance,
+}
+
+#[derive(Encode, Decode, Debug, Default, Eq, PartialEq)]
+pub struct UnstakingState<BlockNumber> {
+    /// The block where the unstaking was initiated
+    pub started_at_block: BlockNumber,
+
+    /// Whether unstaking is in active, or conversely paused state.
+    /// Blocks are only counted towards unstaking period when active.
+    pub is_active: bool,
+
+    /// The number blocks which must be finalised while in the active period before the unstaking is finished
+    pub blocks_remaining_in_active_period_for_unstaking: BlockNumber,
+}
+
+#[derive(Encode, Decode, Debug, Eq, PartialEq)]
+pub enum StakedStatus<BlockNumber> {
+    /// Baseline staking status, nothing is happening.
+    Normal,
+
+    /// Unstaking is under way.
+    Unstaking(UnstakingState<BlockNumber>),
+}
+
+impl<BlockNumber> Default for StakedStatus<BlockNumber> {
+    fn default() -> Self {
+        StakedStatus::Normal
+    }
+}
+
+#[derive(Encode, Decode, Debug, Default, Eq, PartialEq)]
+pub struct StakedState<BlockNumber, Balance, SlashId: Ord> {
+    /// Total amount of funds at stake.
+    pub staked_amount: Balance,
+
+    /// Status of the staking.
+    pub staked_status: StakedStatus<BlockNumber>,
+
+    /// SlashId to use for next Slash that is initiated.
+    /// Will be incremented by one after adding a new Slash.
+    pub next_slash_id: SlashId,
+
+    /// All ongoing slashing.
+    pub ongoing_slashes: BTreeMap<SlashId, Slash<BlockNumber, Balance>>,
+}
+
+impl<BlockNumber, Balance, SlashId> StakedState<BlockNumber, Balance, SlashId>
+where
+    BlockNumber: SimpleArithmetic + Copy,
+    Balance: SimpleArithmetic + Copy,
+    SlashId: Ord + Copy,
+{
+    /// Iterates over all ongoing slashes and decrements blocks_remaining_in_active_period_for_slashing of active slashes (advancing the timer).
+    /// Returns the the number of slashes that were found to be active and had their timer advanced.
+    fn advance_slashing_timer(&mut self) -> usize {
+        self.ongoing_slashes
+            .iter_mut()
+            .fold(0, |slashes_updated, (_slash_id, slash)| {
```

Comment:
I do not think its advisable to mutate inside of _fold_, it breaks the semantics.

Reply:
used in a for..in loop instead